zrml/neo-swaps/src/math.rs

```diff
@@ -606,4 +712,66 @@ mod tests {
             exp(FixedType::checked_from_num(EXP_OVERFLOW_THRESHOLD).unwrap(), neg).unwrap();
         assert_eq!(result, expected);
     }
+
+    #[test_case(_9_10, _10, _1_10, 219722457734)] // Large price shift
+    #[test_case(_4_10, _10, _3_10, 15415067983)] // Small price shift
+    #[test_case(_3_10, _10, _4_10, 0)] // Zero buy amount
+    #[test_case(_4_10, _10, _4_10, 0)] // Zero buy amount
+    #[test_case(_1_100, _10, _1_2, 0)] // Very small until
+    #[test_case(1_2, _10, 1_100, 0)] // Very small spot_price
+    fn calculate_buy_amount_until_works(
+        until: MockBalance,
+        liquidity: MockBalance,
+        spot_price: MockBalance,
+        expected: MockBalance,
+    ) {
+        assert_eq!(
+            MockMath::calculate_buy_amount_until(until, liquidity, spot_price).unwrap(),
+            expected
+        );
+    }
+
+    #[test_case(_100, _10, _1_2)] // Very large until
+    #[test_case(_1_2, _10, _100)] // Very large spot_price
+    fn calculate_buy_amount_until_throws_math_error(
+        until: MockBalance,
+        liquidity: MockBalance,
+        spot_price: MockBalance,
+    ) {
+        assert_err!(
+            MockMath::calculate_buy_amount_until(until, liquidity, spot_price),
+            Error::<MockRuntime>::MathError
+        );
+    }
+
+    #[test_case(_1_10, _10, _9_10, 439444915467)] // Large price shift
+    #[test_case(_1_10, _10, _2_10, 81093021622)] // Small price shift
+    #[test_case(_2_10, _10, _1_10, 0)] // Zero sell amount
+    #[test_case(_1_10, _10, _1_10, 0)] // Zero sell amount
+    #[test_case(_1_100, _10, _1_2, 459511985013)] // Very small until
+    #[test_case(1_2, _10, 1_100, 451815891780)] // Very small spot_price
+    fn calculate_sell_amount_until_fixed_works(
+        until: MockBalance,
+        liquidity: MockBalance,
+        spot_price: MockBalance,
+        expected: MockBalance,
+    ) {
+        assert_eq!(
+            MockMath::calculate_sell_amount_until(until, liquidity, spot_price).unwrap(),
+            expected
+        );
+    }
+
+    #[test_case(_100, _10, _1_2)] // Very large until
+    #[test_case(_1_2, _10, _100)] // Very large spot_price
+    fn calculate_sell_amount_until_throws_math_error(
+        until: MockBalance,
+        liquidity: MockBalance,
+        spot_price: MockBalance,
+    ) {
+        assert_err!(
+            MockMath::calculate_sell_amount_until(until, liquidity, spot_price),
+            Error::<MockRuntime>::MathError
+        );
+    }
```

Comment:
I've added a couple of cases that I think are missing

```suggestion
    #[test_case(_9_10, _10, _1_10, 219722457734)] // Large price shift
    #[test_case(_4_10, _10, _3_10, 15415067983)] // Small price shift
    #[test_case(_3_10, _10, _4_10, 0)] // Zero buy amount
    #[test_case(_4_10, _10, _4_10, 0)] // Zero buy amount
    #[test_case(MAX_SPOT_PRICE, 188_739_165_817, MIN_SPOT_PRICE, 999_053_937_034; "leap_up")]
    #[test_case(MIN_SPOT_PRICE, _10, MAX_SPOT_PRICE, 0; "leap_down")]
    #[test_case(
        MIN_SPOT_PRICE + 100_000,
        132_117_416_072,
        MIN_SPOT_PRICE,
        1_327_820;
        "step_up_low"
    )]
    #[test_case(
        MAX_SPOT_PRICE,
        11_324_349_949,
        MAX_SPOT_PRICE - 100_000,
        22_626_081;
        "step_up_high"
    )]
    #[test_case(MIN_SPOT_PRICE, _1, MIN_SPOT_PRICE + 100_000, 0; "step_down_low")]
    #[test_case(MAX_SPOT_PRICE - 100_000, _1, MAX_SPOT_PRICE, 0; "step_down_high")]
    fn calculate_buy_amount_until_works(
        until: MockBalance,
        liquidity: MockBalance,
        spot_price: MockBalance,
        expected: MockBalance,
    ) {
        assert_eq!(
            MockMath::calculate_buy_amount_until(until, liquidity, spot_price).unwrap(),
            expected
        );
    }

    #[test_case(_1, _10, _1_2; "until too large")]
    #[test_case(_1_2, _10, _1; "spot price too large")]
    #[test_case(u128::MAX, _10, _1_2; "until limit")]
    #[test_case(_3_4, u128::MAX, _1_2; "liquidity limit")]
    #[test_case(_3_4, _10, u128::MAX; "spot price limit")]
    fn calculate_buy_amount_until_throws_math_error(
        until: MockBalance,
        liquidity: MockBalance,
        spot_price: MockBalance,
    ) {
        assert_err!(
            MockMath::calculate_buy_amount_until(until, liquidity, spot_price),
            Error::<MockRuntime>::MathError
        );
    }

    #[test_case(_1_10, _10, _9_10, 439444915467)] // Large price shift
    #[test_case(_1_10, _10, _2_10, 81093021622)] // Small price shift
    #[test_case(_2_10, _10, _1_10, 0)] // Zero sell amount
    #[test_case(_1_10, _10, _1_10, 0)] // Zero sell amount
    #[test_case(_1_100, _10, _1_2, 459511985013)] // Very small until
    #[test_case(1_2, _10, 1_100, 451815891780)] // Very small spot_price
    #[test_case(MAX_SPOT_PRICE, 188_739_165_817, MIN_SPOT_PRICE, 0; "leap_up")]
    #[test_case(MIN_SPOT_PRICE, 11_324_349_949, MAX_SPOT_PRICE, 119_886_472_444; "leap_down")]
    #[test_case(MIN_SPOT_PRICE + 100_000, 132_117_416_072, MIN_SPOT_PRICE, 0; "step_up_low")]
    #[test_case(MAX_SPOT_PRICE, 11_324_349_949, MAX_SPOT_PRICE - 100_000, 0; "step_up_high")]
    #[test_case(
        MIN_SPOT_PRICE,
        186_922_262_798,
        MIN_SPOT_PRICE + 100_000,
        375_349_804;
        "step_down_low"
    )]
    #[test_case(
        MAX_SPOT_PRICE - 100_000,
        43_410_008_138,
        MAX_SPOT_PRICE,
        87_169_596;
        "step_down_high"
    )]
    fn calculate_sell_amount_until_fixed_works(
        until: MockBalance,
        liquidity: MockBalance,
        spot_price: MockBalance,
        expected: MockBalance,
    ) {
        assert_eq!(
            MockMath::calculate_sell_amount_until(until, liquidity, spot_price).unwrap(),
            expected
        );
    }

    #[test_case(0, _10, _1_2; "until too small")]
    #[test_case(_1_2, _10, _1; "spot price too large")]
    #[test_case(u128::MAX, _10, _3_4; "until limit")]
    #[test_case(_1_2, u128::MAX, _3_4; "liquidity limit")]
    #[test_case(_1_2, _10, u128::MAX; "spot price limit")]
    fn calculate_sell_amount_until_throws_math_error(
        until: MockBalance,
        liquidity: MockBalance,
        spot_price: MockBalance,
    ) {
        assert_err!(
            MockMath::calculate_sell_amount_until(until, liquidity, spot_price),
            Error::<MockRuntime>::MathError
        );
    }
```

Reply:
f31beee27350cb6f5d6b8c8c8fdae168bce97698